### PR TITLE
Introduce `default_encoding` parameter to [set|autodetect] the encoding if the charset is missing from the headers

### PR DESCRIPTION
**Pull request by deedy5**
## PR summary:

1. Added `default_encoding` parameter
2. Updated github actions
_____

### Using the default encoding

When requests are made to a site without explicit character set information from the server, but the encoding is known, it's advisable to explicitly set the default encoding.

```python
from curl_cffi import requests

resp = requests.get(..., default_encoding="shift-jis")
print(response.encoding)  # Shows charset given in the Content-Type charset, or else "shift-jis".
print(response.text)  # The text decoded with the Content-Type charset, or using "shift-jis".
```

**Using character set auto-detection**

When the server doesn't provide character set information reliably, and the encoding is unknown, enabling auto-detection allows for a best-effort guess when converting bytes to text. To activate auto-detection, set the default_encoding parameter to a function that takes input bytes and returns the appropriate character set for decoding.

Let's take a look at autodetection using [`charset-normalizer`](https://github.com/Ousret/charset_normalizer)

```python
from curl_cffi import requests
from charset_normalizer import detect

def autodetect(content):
    return detect(content).get("encoding")

# Using a session with character-set autodetection enabled.
s = requests.Session(default_encoding=autodetect)
resp = s.get(...)
print(resp.encoding)  # Shows charset given in the Content-Type charset, or else autodetect() result.
print(resp.text)  # The text decoded with the Content-Type charset, or encoding, detected via autodetect.
```

### CODE REVIEW

**Comment by T-256** [comment]
fyi httpx uses utf-8 with `errors=replace` as default fallback. it uses charset_normalizer as optional feature.

**Comment by deedy5** [comment]
`requests` - fallback to auto-detected encoding
https://github.com/psf/requests/blob/8dd3b26bf59808de24fd654699f592abf6de581e/src/requests/models.py#L924
https://github.com/psf/requests/blob/8dd3b26bf59808de24fd654699f592abf6de581e/src/requests/models.py#L790

**Comment by T-256** [comment]
> `requests` - fallback to auto-detected encoding

Needs decision. (I prefer httpx way, to not force new dependency.)

**Comment by deedy5** [comment]
httpx way: https://github.com/encode/httpx/blob/392dbe45f086d0877bd288c5d68abf860653b680/httpx/_models.py#L587

**Comment by deedy5** [comment]
httpx PR with useful information
https://github.com/encode/httpx/pull/2165/

File: curl_cffi/requests/session.py [hunk 1]
```diff
@@ -279,10 +281,7 @@ def _set_curl_options(
         c = curl
 
         # method
-        if method == "POST":
-            c.setopt(CurlOpt.POST, 1)
-        elif method != "GET":
-            c.setopt(CurlOpt.CUSTOMREQUEST, method.encode())
+        c.setopt(CurlOpt.CUSTOMREQUEST, method.encode())
```
Comment (yifeikong) [inline curl_cffi/requests/session.py:284]: `CURLOPT_POST` is not the same as `CURLOPT_CUSTOMREQUEST` with "POST", see #135

**Comment by yifeikong** [comment]
I also prefer to add this as an optional feature.

File: curl_cffi/requests/session.py (hunk 1 above)
Comment (deedy5) [reply curl_cffi/requests/session.py]: Fixed https://github.com/yifeikong/curl_cffi/pull/284/commits/1aa8c23c67408012b0f3c538e80ac0408adb71e3

**Comment by deedy5** [comment]
@yifeikong :checkered_flag: 
Ignore the check error in `Build, test and release / Build bdist wheels and test (macos-14) (pull_request)`.
Just run the checks later, it's some bugs on the site.

**Comment by deedy5** [comment]
@yifeikong Update the documentation where you see fit, I've added a description in the first post.

**Review by T-256 — APPROVED** [review]
Nice, Thank you!

File: .github/workflows/build-and-test.yaml [hunk 2]
```diff
@@ -31,15 +31,15 @@ jobs:
     name: Build sdist wheel
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v3
+      - uses: actions/checkout@v4
       - run: |
           make preprocess
           pipx run build --sdist
-      - uses: actions/upload-artifact@v3
+      - uses: actions/upload-artifact@v4
```
Comment (bashonly) [inline .github/workflows/build-and-test.yaml:38]: Just curious: why are you making all these unrelated changes in this PR? you know v4 of `actions/upload-artifact` has [breaking changes](https://github.com/actions/upload-artifact/blob/main/docs/MIGRATION.md), right?

Comment (deedy5) [reply .github/workflows/build-and-test.yaml]: Thanks for pointing out the changes to v4 actions/upload-artifact. I've been experimenting with different versions and configurations to see why GitHub Actions was causing errors. The commits you saw were part of those experiments. I really appreciate your feedback and hope you and also continue to help with development.

**Comment by deedy5** [comment]
The problem occurred when installing pipx. 
But as it turned out, it is already installed https://github.com/pypa/cibuildwheel#v2170
So I removed that action and all works.

**Comment by deedy5** [comment]
@yifeikong @T-256 :checkered_flag:

**Comment by deedy5** [comment]
Since autoflake, flake8, flake8-bugbear, flake8-pie are removed, the most popular rules for ruff are set.
https://docs.astral.sh/ruff/linter/#rule-selection

**Comment by coletdjnz** [comment]
I recommend splitting out unrelated changes into another PR, as it helps with review (and for downstream users to audit)

**Comment by deedy5** [comment]
Split the PR into several, in this one I left only adding the `default_encoder` parameter and updating the github actions to remove the bug in the tests.

**Comment by yifeikong** [comment]
Generally, I think the `encoding/charset` properies should be kept, other wise it would be a breaking change, which should be introduced only when the major version bumps.
